p4-16/spec/P4-16-spec.mdk

```diff
@@ -3489,6 +3489,11 @@ Bit-strings also support the following operations:
   The effect of this statement is to set bits `H` through `L` (inclusive of
   both) of `e` to the bit-pattern represented by `x`, and leaves all other bits
   of `e` unchanged.  A slice of an unsigned integer is an unsigned integer.
+  A slice may also be specified as `[L+:W]` where L is the lowest bit of the slice
+  and W is the width of the slice.  In this case, only W must be a non-negative
+  local compile-time known numeric value.  L must also be a numeric value and
+  some architectures may require it to be compile-time known.  Some architectures
+  may allow variable indexing.
```

Comment:
This wording currently allows `L` to be negative (e.g. either negative `int` literal, or `int<N>` expression which can have a negative value). A way out would be to say that its type must be either `int`, in which cast it must be non-negative, or `bit<N>`-expression. I think it would make sense to restrict this, as slices on negative bit index don't make sense.

Reply:
Yes, this is tricky.  I added some verbiage to require non-negative in-range values, but only require an error message for constants, as it may be tricky to do that for variable values, and we don't have any way of dealing with it at runtime.